File: src/harness_core/event_schema.py
"""Minimal event.v1 schema validation for Stage 1 Day 1."""

from __future__ import annotations

import hashlib
import json
from typing import Any

from .errors import SchemaViolationError

REQUIRED_FIELDS = {
    "event_id",
    "schema_version",
    "event_type",
    "timestamp",
    "producer",
    "correlation",
    "severity",
    "payload",
    "idempotency_key",
    "parent_event_id",
}

REQUIRED_PRODUCER_FIELDS = {"component_id", "component_type"}
VALID_SEVERITIES = {"info", "warn", "error"}
IDEMPOTENCY_HASH_EXCLUDED_FIELDS = {"event_id", "timestamp"}


def validate_event(event: dict[str, Any]) -> None:
    """Validate the minimal event.v1 contract.

    This intentionally avoids implementing the full future validator suite.
    """
    if not isinstance(event, dict):
        raise SchemaViolationError("event must be a JSON object")

    missing = sorted(REQUIRED_FIELDS - event.keys())
    if missing:
        raise SchemaViolationError(f"missing required field(s): {', '.join(missing)}")

    _require_string(event, "event_id")
    _require_string(event, "event_type")
    _require_string(event, "timestamp")
    _require_string(event, "idempotency_key")

    if event["schema_version"] != "event.v1":
        raise SchemaViolationError("schema_version must be event.v1")

    if event["severity"] not in VALID_SEVERITIES:
        raise SchemaViolationError("severity must be one of: error, info, warn")

    if not isinstance(event["producer"], dict):
        raise SchemaViolationError("producer must be an object")
    missing_producer = sorted(REQUIRED_PRODUCER_FIELDS - event["producer"].keys())
    if missing_producer:
        raise SchemaViolationError(
            f"producer missing required field(s): {', '.join(missing_producer)}"
        )
    _require_string(event["producer"], "component_id", prefix="producer.")
    _require_string(event["producer"], "component_type", prefix="producer.")

    if not isinstance(event["correlation"], dict):
        raise SchemaViolationError("correlation must be an object")

    if not isinstance(event["payload"], dict):
        raise SchemaViolationError("payload must be an object")

    parent_event_id = event["parent_event_id"]
    if parent_event_id is not None and not isinstance(parent_event_id, str):
        raise SchemaViolationError("parent_event_id must be a string or null")


def stable_idempotency_hash(event: dict[str, Any]) -> str:
    """Hash semantic event content, excluding volatile identity fields."""
    semantic_event = {
        key: value
        for key, value in event.items()
        if key not in IDEMPOTENCY_HASH_EXCLUDED_FIELDS
    }
    canonical = json.dumps(semantic_event, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def canonical_event_json(event: dict[str, Any]) -> str:
    """Serialize an event as one canonical JSON object."""
    return json.dumps(event, sort_keys=True, separators=(",", ":"))


def _require_string(mapping: dict[str, Any], key: str, prefix: str = "") -> None:
    if not isinstance(mapping[key], str) or not mapping[key]:
        raise SchemaViolationError(f"{prefix}{key} must be a non-empty string")

File: engine/src/event_schema.rs
//! Reference-only module: event.v1 schema validation and idempotency hashing.
//!
//! Retained for parity with wire_contract/v1 event schemas and golden fixtures.
//! The active runtime uses `LocalProductStore` audit events instead.

use std::collections::HashSet;
use std::sync::LazyLock;

use serde_json::{Map, Value};
use sha2::{Digest, Sha256};

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

pub static REQUIRED_FIELDS: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    let mut s = HashSet::with_capacity(10);
    s.insert("event_id");
    s.insert("schema_version");
    s.insert("event_type");
    s.insert("timestamp");
    s.insert("producer");
    s.insert("correlation");
    s.insert("severity");
    s.insert("payload");
    s.insert("idempotency_key");
    s.insert("parent_event_id");
    s
});

pub static REQUIRED_PRODUCER_FIELDS: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    let mut s = HashSet::with_capacity(2);
    s.insert("component_id");
    s.insert("component_type");
    s
});

pub static VALID_SEVERITIES: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    let mut s = HashSet::with_capacity(3);
    s.insert("info");
    s.insert("warn");
    s.insert("error");
    s
});

pub static IDEMPOTENCY_HASH_EXCLUDED_FIELDS: LazyLock<HashSet<&'static str>> =
    LazyLock::new(|| {
        let mut s = HashSet::with_capacity(2);
        s.insert("event_id");
        s.insert("timestamp");
        s
    });

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct SchemaViolationError(pub String);

impl std::fmt::Display for SchemaViolationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for SchemaViolationError {}

// ---------------------------------------------------------------------------
// Canonical JSON (sorted keys, compact separators — mirrors Python json.dumps)
// ---------------------------------------------------------------------------

fn canonical_json(value: &Value) -> String {
    match value {
        Value::Object(map) => {
            let mut pairs: Vec<_> = map.iter().collect();
            pairs.sort_by_key(|(k, _)| k.as_str());
            let inner: String = pairs
                .iter()
                .map(|(k, v)| format!("\"{}\":{}", k, canonical_json(v)))
                .collect::<Vec<_>>()
                .join(",");
            format!("{{{}}}", inner)
        }
        Value::Array(arr) => {
            let inner: String = arr.iter().map(canonical_json).collect::<Vec<_>>().join(",");
            format!("[{}]", inner)
        }
        Value::String(s) => serde_json::to_string(s).unwrap(),
        other => other.to_string(),
    }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

pub fn validate_event(event: &Value) -> Result<(), SchemaViolationError> {
    let object = event
        .as_object()
        .ok_or_else(|| SchemaViolationError("event must be a JSON object".to_string()))?;

    // Check all required fields are present
    let missing: Vec<&str> = REQUIRED_FIELDS
        .iter()
        .copied()
        .filter(|field| !object.contains_key(*field))
        .collect();
    if !missing.is_empty() {
        let mut sorted = missing;
        sorted.sort();
        return Err(SchemaViolationError(format!(
            "missing required field(s): {}",
            sorted.join(", ")
        )));
    }

    // Non-empty string checks
    require_non_empty_string(object, "event_id")?;
    require_non_empty_string(object, "event_type")?;
    require_non_empty_string(object, "timestamp")?;
    require_non_empty_string(object, "idempotency_key")?;

    // schema_version must be exactly "event.v1"
    if object.get("schema_version").and_then(Value::as_str) != Some("event.v1") {
        return Err(SchemaViolationError(
            "schema_version must be event.v1".to_string(),
        ));
    }

    // severity must be one of valid values
    match object.get("severity").and_then(Value::as_str) {
        Some(s) if VALID_SEVERITIES.contains(s) => {}
        _ => {
            return Err(SchemaViolationError(
                "severity must be one of: error, info, warn".to_string(),
            ))
        }
    }

    // producer must be an object with required sub-fields
    let producer = object
        .get("producer")
        .and_then(Value::as_object)
        .ok_or_else(|| SchemaViolationError("producer must be an object".to_string()))?;
    let missing_producer: Vec<&str> = REQUIRED_PRODUCER_FIELDS
        .iter()
        .copied()
        .filter(|field| !producer.contains_key(*field))
        .collect();
    if !missing_producer.is_empty() {
        let mut sorted = missing_producer;
        sorted.sort();
        return Err(SchemaViolationError(format!(
            "producer missing required field(s): {}",
            sorted.join(", ")
        )));
    }
    require_non_empty_string(producer, "component_id")?;
    require_non_empty_string(producer, "component_type")?;

    // correlation must be an object
    if !object.get("correlation").is_some_and(Value::is_object) {
        return Err(SchemaViolationError(
            "correlation must be an object".to_string(),
        ));
    }

    // payload must be an object
    if !object.get("payload").is_some_and(Value::is_object) {
        return Err(SchemaViolationError(
            "payload must be an object".to_string(),
        ));
    }

    // parent_event_id must be null or string
    match object.get("parent_event_id") {
        Some(Value::Null) | Some(Value::String(_)) => {}
        _ => {
            return Err(SchemaViolationError(
                "parent_event_id must be a string or null".to_string(),
            ))
        }
    }

    Ok(())
}

pub fn canonical_event_json(event: &Value) -> Result<String, SchemaViolationError> {
    Ok(canonical_json(event))
}

pub fn stable_idempotency_hash(event: &Value) -> Result<String, SchemaViolationError> {
    let object = event
        .as_object()
        .ok_or_else(|| SchemaViolationError("event must be a JSON object".to_string()))?;

    let mut filtered = Map::with_capacity(object.len());
    for (key, value) in object {
        if !IDEMPOTENCY_HASH_EXCLUDED_FIELDS.contains(key.as_str()) {
            filtered.insert(key.clone(), value.clone());
        }
    }

    let canonical = canonical_json(&Value::Object(filtered));
    let digest = Sha256::digest(canonical.as_bytes());
    Ok(to_hex(&digest))
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn require_non_empty_string(
    object: &Map<String, Value>,
    key: &str,
) -> Result<(), SchemaViolationError> {
    match object.get(key).and_then(Value::as_str) {
        Some(s) if !s.is_empty() => Ok(()),
        _ => Err(SchemaViolationError(format!(
            "{} must be a non-empty string",
            key
        ))),
    }
}

fn to_hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}
